libs/depgraph/README.md

```diff
@@ -0,0 +1,284 @@
+                Type: "Linux interface", // can also use LinuxInterface{}.Type()
+                Name: route.via, // can also use LinuxInterface{name: route.via}.Name()
+            },
+            Description: "Route requires outgoing interface to be configured first",
+        },
+    }
+}
+
+// Other Item methods for LinuxRoute are omitted.
+```
+
+Once all needed `Item` implementations are available, you are ready to build a dependency
+graph to model some kind of system state.
+Graph with some initial content is created using `New()`:
+
+```go
+import "github.com/lf-edge/eve/libs/depgraph"
+
+// using LinuxRoute and LinuxInterface defined above
+
+g := depgraph.New(depgraph.InitArgs{
+    Name:        "MyGraph",
+    Description: "This is my example graph",
+    Items:       []Item{
+        LinuxRoute{via: "eth0", dst: "10.10.0.0/16", metric: 10},
+        LinuxRoute{via: "eth0", dst: "192.168.16.0/24", metric: 10},
+        LinuxInterface{name: "eth0"},
+    },
+})
+```
+
+Graph will automatically create one node for each item and will build edges based on
+dependencies of these items. In the example above, there will be edge from both routes
+pointing to the `eth0` interface.
+`InitArgs` can also contain the initial content of subgraphs.
+
+A single Node is manipulated using `Node()`, `PutNode()` and `DelNode()` methods:
+
+```go
+import (
+    "fmt"
+    "github.com/lf-edge/eve/libs/depgraph"
+)
+
+// Add new Linux route.
+// Note that `PutNode()` takes `Node` pointer so that a reference to the graph can be
+// added into the Node (this is needed for `Node.AsGraph()`).
+n := depgraph.Node{Item: LinuxRoute{via: "eth0", dst: "10.20.0.0/16", metric: 100}}
+g.PutNode(&n)
+
+// Update the node.
+n.metric = 50
+g.PutNode(&n)
+
+// Delete the node.
+g.DelNode(n.ID())
+
+// Read graph node.
+n, found := g.Node(depgraph.NodeIDFor(LinuxInterface{name: "eth0"}))
+if found {
+    fmt.Printf("Linux interface graph node: %+v\n", n)
+}
+```
+
+Note that Node is not only a container for Item, but is also able to store some additional
+information which may be useful for state reconciliation purposes.
+
+To iterate all nodes in the graph:
+
+```go
+inclSubGraphs := true
+iter := g.Nodes(inclSubGraphs)
+for iter.Next() {
+    n := iter.Node()
+    fmt.Printf("Graph node: %+v\n", n)
+}
+```
+
+To iterate all edges originating from a node:
+
+```go
+iter := g.OutgoingEdges(n.ID())
+for iter.Next() {
+    e := iter.Edge()
+    fmt.Printf("Edge from %s to %s for dep: %+v\n",
+        e.FromNode, e.ToNode, e.Dependency)
+}
+```
+
+Lastly, sub-graphs are manipulated using `SubGraph()`, `PutSubGraph()`, `DelSubGraph`
+and `EditSubGraph` methods:
+
+```go
+item1 := LinuxInterface{name: "eth0.1"}
+item2 := LinuxInterface{name: "eth0.2"}
+subG := depgraph.New(depgraph.InitArgs{
+    Name:        "MySubGraph",
+    Description: "This is my example sub-graph",
+    Items:       []Item{item1, item2},
+})
+
+// Add new subgraph
+g.PutSubGraph(subG)
```

Comment:
I found this somewhat confusing. How can I out a new graph as a subgraph to an existing graph with just `Put`? Don't I need to make it a child of _something_, some existing node in the graph?

Reply:
In this example `subG` will become child of `g`. If you then want to add subgraph into `subG` (nested from `g` point of view), then you will do something like `subG.PutSubGraph(nestedSubG)`.
Here I drew an example of a depgraph with subgraphs: https://drive.google.com/file/d/1VPZl6W0ec8tOvtuiVZdV8Y-hmA6Vvdrb/view?usp=sharing

Reply:
Got it. OK, that makes sense. Actually, it makes a lot of sense.